lib/tables.c

```diff
@@ -2881,374 +2874,297 @@ simplifier_free(simplifier_t *self)
+        if (ret != 0) {
+            goto out;
+        }
+        prev_right = right;
     }
-    if (defrag_required) {
-        ret = simplifier_defrag_segment_chain(self, input_id);
+    /* Check for errors occuring in the loop condition */
+    if (ret != 0) {
+        goto out;
+    }
+    if (is_sample && prev_right != self->sequence_length) {
+        /* If a trailing gap exists in the sample ancestry, fill it in. */
+        ret = simplifier_add_ancestry(self, input_id, prev_right,
+                self->sequence_length, output_id);
         if (ret != 0) {
             goto out;
         }
     }
-    ret = simplifier_flush_edges(self);
+    ret = simplifier_flush_edges(self, output_id);
 out:
     return ret;
 }
 
+static int WARN_UNUSED
+simplifier_process_parent_edges(simplifier_t *self, node_id_t parent, size_t start,
+        size_t end)
+{
+    int ret = 0;
+    size_t j;
+    simplify_segment_t *x;
+    node_id_t child;
+    double left, right;
+
+    /* Go through the edges and queue up ancestry segments for processing. */
+    self->segment_queue_size = 0;
+    for (j = start; j < end; j++) {
+        assert(parent == self->input_edges.parent[j]);
+        child = self->input_edges.child[j];
+        left = self->input_edges.left[j];
+        right = self->input_edges.right[j];
+        for (x = self->ancestor_map_head[child]; x != NULL; x = x->next) {
+            if (x->right > left && right > x->left) {
+                ret = simplifier_enqueue_segment(self,
+                        MSP_MAX(x->left, left), MSP_MIN(x->right, right), x->node);
+                if (ret != 0) {
+                    goto out;
+                }
+            }
+        }
+    }
+    /* We can now merge the ancestral segments for the parent */
+    ret = simplifier_merge_ancestors(self, parent);
+    if (ret != 0) {
+        goto out;
+    }
+out:
+    return ret;
+}
+
+static int WARN_UNUSED
+simplifier_map_mutation_nodes(simplifier_t *self)
+{
+    int ret = 0;
+    simplify_segment_t *seg;
+    mutation_id_list_t *m_node;
+    size_t input_node;
+    site_id_t site;
+    double position;
+
+    for (input_node = 0; input_node < self->input_nodes.num_rows; input_node++) {
+        seg = self->ancestor_map_head[input_node];
+        m_node = self->node_mutation_list_map_head[input_node];
+        /* Co-iterate over the segments and mutations; mutations must be listed
+         * in increasing order of site position */
+        while (seg != NULL && m_node != NULL) {
+            site = self->input_mutations.site[m_node->mutation];
+            position = self->input_sites.position[site];
+            if (seg->left <= position && position < seg->right) {
+                self->mutation_node_map[m_node->mutation] = seg->node;
+                m_node = m_node->next;
+            } else if (position >= seg->right) {
+                seg = seg->next;
+            } else {
+                assert(position < seg->left);
```

Comment:
I don't see how this could ever be triggered, as the previous conditions cover `position >= seg->left`.